functorch/_src/aot_autograd.py

```diff
@@ -372,12 +376,22 @@ class AOTConfig:
     num_params_buffers: int
 
 
-def aot_dispatch_base(flat_fn, flat_args: List[Tensor], aot_config: AOTConfig):
+
+def aot_dispatch_base(flat_fn, flat_args: List[Tensor], aot_config: AOTConfig, shape_env):
+    if shape_env:
+        pre_dispatch_guards = shape_env.guards
```

Comment:
I just fundamentally don't understand the guard strategy here. If I want a two stage cache, it would make far more sense to me for ShapeEnv to have a built-in notion of tiering, instead of manually doing staging like is being done here. In fact, I might just suggest cloning the shape env entirely, because when we do AOTAutograd lowering we may reveal equalities that can simplify the original guards from dynamo... but those simplifications may not be sound in general.

TBH, I still kind of think we should do a one level cache first, it's conceptually simpler, and dynamo tracing isn't the bottleneck on this side of the stack (aot autograd is.)

Reply:
Discussed offline, a lot, and updated the comments here to match.